Split Redis docs (#11246)
gravitational/teleport: docs/pages/database-access/guides/redis-cluster.mdx

description: How to configure Teleport Database Access with Redis Cluster.
---

<Details
  title="Version warning"
  opened={true}
  scope={["oss", "enterprise"]}
  scopeOnly={true}
  min="9.0"
>
  Database access for Redis is available starting from Teleport `9.0`.
</Details>

If you want to configure Redis Standalone guide is available [here](redis.mdx).

This guide will help you to:

- Install and configure Teleport.
- Configure mutual TLS authentication between Teleport and Redis Cluster.
- Connect to Redis through Teleport.

## Prerequisites

- Teleport version `9.0` or newer.
- Redis version `6.0` or newer.
- `redis-cli` installed and added to your system's `PATH` environment variable.

<Admonition type="note" title="Note">
  Redis `7.0` and RESP3 (REdis Serialization Protocol) are currently not supported.
</Admonition>

## Step 1/6. Set up Teleport Auth and Proxy

(!docs/pages/includes/database-access/start-auth-proxy.mdx!)

## Step 2/6. Create a Teleport user

(!docs/pages/includes/database-access/create-user.mdx!)

## Step 3/6. Create Redis users

(!docs/pages/includes/database-access/redis-create-users.mdx!)

## Step 4/6. Set up mutual TLS

> (!docs/pages/includes/database-access/tctl-auth-sign.mdx!)

Reviewer: I think there should be a clearer transition between the use of the `tctl-auth-sign.mdx` partial and the `### Create the secrets` section. Added a suggestion:

```suggestion
(!docs/pages/includes/database-access/tctl-auth-sign.mdx!)

We will show you how to use the `tctl auth sign` command below.
```

> ### Create the secrets

Reviewer: I think we can remove this heading since it's the only H3-level heading in the "Set up mutual TLS" section.

```suggestion
```

export CERTS_DIR=/path/to/certs/
export IP1=10.0.0.1 # update with the real node 1 IP
export IP2=10.0.0.2 # update with the real node 2 IP
export IP3=10.0.0.3 # update with the real node 3 IP
export IP4=10.0.0.4 # update with the real node 4 IP
export IP5=10.0.0.5 # update with the real node 5 IP
export IP6=10.0.0.6 # update with the real node 6 IP
redis-cli --user alice --cluster-replicas 1 --tls --cluster-yes \
--cluster create ${IP1}:7001 ${IP2}:7002 ${IP3}:7003 ${IP4}:7004 ${IP5}:7005 ${IP6}:7006 \
--cacert ${CERTS_DIR}/server.cas --key ${CERTS_DIR}/server.key --cert ${CERTS_DIR}/server.crt
```

## Step 6/6. Connect

To enable Redis cluster mode in Teleport, add the `mode=cluster` parameter to the connection URI in
your Teleport Database Service config file.
```yaml
databases:
  - name: "redis-cluster"
    uri: "rediss://redis.example.com:6379?mode=cluster"
```

(!docs/pages/includes/database-access/redis-connect.mdx!)

### Supported Redis Cluster commands

Redis in cluster mode doesn't support below commands. If one of the listed commands above is called Teleport
returns the <nobr>`ERR Teleport: command not supported`</nobr> error.

> - `ACL`

Reviewer 2: Can this long list be hidden under collapsible section? I think we have a "details" component for this.

- `ASKING`
- `CLIENT`
- `CLUSTER`
- `CONFIG`
- `DEBUG`
- `EXEC`
- `HELLO`
- `INFO`
- `LATENCY`
- `MEMORY`
- `MIGRATE`
- `MODULE`
- `MONITOR`
- `MULTI`
- `PFDEBUG`
- `PFSELFTEST`
- `PSUBSCRIBE`
- `PSYNC`
- `PUNSUBSCRIBE`
- `PUNSUBSCRIBE`
- `READONLY`
- `READWRITE`
- `REPLCONF`
- `REPLICAOF`
- `ROLE`
- `SCAN`
- `SCRIPT DEBUG`
- `SCRIPT KILL`
- `SHUTDOWN`
- `SLAVEOF`
- `SLOWLOG`
- `SSUBSCRIBE`
- `SUNSUBSCRIBE`
- `SYNC`
- `TIME`
- `WAIT`
- `WATCH`

> Teleport also adds additional processing to:

Reviewer: ```suggestion
Teleport conducts additional processing on the following commands before 
communicating with Redis Cluster:
```
or something more specific/less repetitive than "adds additional"

## Step 5/6. Create cluster

Use the following command to create the cluster. Please note `redis-cli --cluster create` accepts only IP addresses.
```sh
export REDISCLI_AUTH=STRONG_GENERATED_PASSWORD
export CERTS_DIR=/path/to/certs/
export IP1=10.0.0.1 # update with the real node 1 IP
export IP2=10.0.0.2 # update with the real node 2 IP
export IP3=10.0.0.3 # update with the real node 3 IP
export IP4=10.0.0.4 # update with the real node 4 IP
export IP5=10.0.0.5 # update with the real node 5 IP
export IP6=10.0.0.6 # update with the real node 6 IP
redis-cli --user alice --cluster-replicas 1 --tls --cluster-yes \
--cluster create ${IP1}:7001 ${IP2}:7002 ${IP3}:7003 ${IP4}:7004 ${IP5}:7005 ${IP6}:7006 \
--cacert ${CERTS_DIR}/server.cas --key ${CERTS_DIR}/server.key --cert ${CERTS_DIR}/server.crt
```

## Step 6/6. Connect

To enable Redis cluster mode in Teleport, add the `mode=cluster` parameter to the connection URI in
your Teleport Database Service config file.
```yaml
databases:
  - name: "redis-cluster"
    uri: "rediss://redis.example.com:6379?mode=cluster"
```

(!docs/pages/includes/database-access/redis-connect.mdx!)

### Supported Redis Cluster commands

> Redis in cluster mode doesn't support below commands. If one of the listed commands above is called Teleport

Reviewer: ```suggestion
Redis in cluster mode does not support the following commands. If one of the listed commands above is called Teleport
```
